### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by Moosems** [Moosems/bazingo#1]
Features:
- Must allow for tabs (open, close, move)
- Close button that doesn't use file (dynamic images using PIL with the json color and char ✕) and can be binded to a function
- Scrollable
- Double click empty space to open empty Editor
- No flicker

**Comment by Moosems** [comment]
For the close button image use the following code (swap background and foreground with vars in json):
```python
# Dymanic image
from PIL import Image, ImageTk, ImageDraw, ImageFont
from tkinter import Tk, Label

root = Tk()
# Create an image of "✕" in #000000 on a #FFFFFF background (use a font that supports the character)
image = Image.new("RGBA", (17, 17), "#FFFFFF")
draw = ImageDraw.Draw(image)
draw.text((3, 0), "X", fill="#000000", font=ImageFont.truetype("./DIP_module/static/Roboto-regular.ttf", 15))
# Convert the image to a Tkinter PhotoImage
photo_image = ImageTk.PhotoImage(image)
# Create a label with the image
label = Label(root, image=photo_image)
label.pack()
root.mainloop()
```

**Comment by Moosems** [comment]
The source for the Notebook is [here](https://github.com/tcltk/tk/blob/main/library/ttk/notebook.tcl) and [here](https://github.com/tcltk/tk/blob/main/library/ttk/notebook.c), perhaps these can be modified and submitted to the Tcl core so it doesn't have to stay as a package.

**Comment by Moosems** [comment]
See [here](https://github.com/DIP-Editor/DIP-dev/issues/15)

**Comment by Moosems** [comment]
Hey @rdbende, I've been giving this a lot of thought and I'm committed to investing significant time and effort into this project. Before I dive in, I want to make sure I have a clear goal and plan. I would really value your thoughts on what I'm proposing.

My plan is to use a Canvas as the foundation for the TabView widget. This approach has several benefits:

- It allows for smooth animations when moving tabs.
- It's easy to use and flexible.
- It enables scrolling, which is not possible with Frames.
- It eliminates flickering when tabs are added or removed (see [here](https://stackoverflow.com/questions/59114409/moving-a-widget-on-a-canvas-in-python) for a possible implementation).
- It offers more options for customizing other widgets and tagging them.
- It's reasonably efficient.

Here are the success criteria for the widget:

1. The widget must allow for tabs to be opened, closed, moved, renamed, added, and removed.
2. Can make the `TabView` horizontal and vertical
3. There must be a close button that doesn't use a file, but rather dynamically generates an image using PIL, with the tab's background as the image background and the "✕" character in the foreground color set in the ttk styling. The button should be able to be bound to a function. The Canvas is particularly well-suited for this.
4. Tabs should be scrollable.
5. Tabs should be draggable.
6. There should be no flicker when switching tabs, adding tabs, removing tabs, or moving them. This can be achieved by moving the tab instead of forgetting and re-placing it.
7. The widget must be efficient, with no lag when switching tabs or performing other functions. The forget() method can be used to save memory by not rendering the tab internals. To avoid flicker, the new tab can be placed on top of the old one and then the old one immediately forgotten.
8. The widget must support the ability to add images to tabs and group tabs. This can be achieved using buttons that have images, with extra spaces added to the text to accommodate the close button on the opposite side.
9. The widget must have a special styling using Style().lookup("TabView", "option").

I believe that when using the TabView, users should not have to use indexes. Instead, users should have a dictionary of the tabs they've used, with the second value being any extra information they want to keep. 

Here's an example of how the `TabView` might be used:

```python
Copy code
tb = TabView(root)
tb.grid(stick="ew")
tab1 = Tab(tb, widget_to_use_inside, text="Tab 1", close_func=print)
tabs = {
    tab1: [name, path, extra_data]
}
tb.add(tab1)
...
current_tab = tb.current_tab  # (@property)
...
deleted_tab = TabViewEvent.deleted_tab  # Special TabView event classes
info = tabs[deleted_tab]
```
Let me know what you think!

**Comment by rdbende** [comment]
> My plan is to use a Canvas as the foundation for the TabView widget.

Wait. Do you want to draw the entire tabview on a canvas?

> It offers more options for customizing other widgets and tagging them.

Wdym?

> The widget must allow for tabs to be opened, closed, moved, renamed, added, and removed.

What's the difference between opened <-> added and closed <-> removed?

> Can make the TabView horizontal and vertical

Yeah, that wouldn't play nicely with the existing ttk themes.

> The widget must have a special styling using Style().lookup("TabView", "option").

Hmmm, why? At the very least, I think it should inherit the `TNotebook` style, as no theme currently has `TabView`.

> I believe that when using the TabView, users should not have to use indexes. Instead, users should have a dictionary of the tabs they've used, with the second value being any extra information they want to keep.

On the pther hand it would be good to preserve compatibility with tkinter, if this is gonna be a tkinter extension/replacement widget.

The rest LGTM

**Comment by Moosems** [comment]
> Wait. Do you want to draw the entire tabview on a canvas?

No, just the tab bar, the rest is a frame that we grid the tab contents into

> Wdym?

With canvas's you can tag items in ways not possible normally (like dragging and whatnot).

> What's the difference between opened <-> added and closed <-> removed?

Opened is when I change a tab and want to view the contents of another, added is when I add it to the tab bar. Closed and removed are the same though, don't quite know why I wrote that twice.

> Yeah, that wouldn't play nicely with the existing ttk themes.

More customization is better imo with this. Plus, that's why a special style might be useful.

> Hmmm, why? At the very least, I think it should inherit the `TNotebook` style, as no theme currently has `TabView`.

They can use `theme_configure()` to add it pretty easily and if we're going to add vertical as an option then it needs its own anyway.

> On the pther hand it would be good to preserve compatibility with tkinter, if this is gonna be a tkinter extension/replacement widget.

While that would be nice, I want to address the issues of the `Notebook` and allowing things like dragging tabs to change order is a lot harder with indexes.

> The rest LGTM

:D.

**Comment by Moosems** [comment]
@rdbende What do you think of [this](https://wiki.tcl-lang.org/page/Creating+a+BWidget+Widget)?

**Comment by Moosems** [comment]
See Thonny, Custom Tkinter, and txt2'a implementation. Will begin actual work sometime soon.
